### DISCUSSION TITLE
PhidgetSpatial connection to plus

### DISCUSSION THREAD

**Discussion by LisanneVenix** [PlusToolkit/PlusLib#912] [Q&A]
Hi, I am new to Plus. I am trying to connect phidgetSpatial 3/3/3 precision 1044_1 to plus, but I keep getting this error: 

_Connection failed!

Please select another device set and try again!_

I used the example config on the website for phidget but I am not sure what to change in the file. How can I get this started? 
````
<PlusConfiguration version="2.1">
  <DataCollection StartupDelaySec="1.0">
    <DeviceSet 
      Name="PlusServer: PhidgetSpatial MARG sensor" 
      Description="Broadcasting tool tracking data through OpenIGTLink
Keep the device stationary for 2 seconds after connect for accurate gyroscope initialization."
    />
    <Device
      Id="TrackerDevice"
      Type="PhidgetSpatial"
      AhrsAlgorithm="MADGWICK_MARG"
      AhrsAlgorithmGain="1.5"
      TiltSensorWestAxisIndex="1"
      FilteredTiltSensorAhrsAlgorithm="MADGWICK_IMU"
      FilteredTiltSensorAhrsAlgorithmGain="0.01"
      FilteredTiltSensorWestAxisIndex="1"      
      ToolReferenceFrame="Tracker" >
      <DataSources>
        <DataSource Type="Tool" Id="OrientationSensor" PortName="OrientationSensor" />    
        <DataSource Type="Tool" Id="TiltSensor" PortName="TiltSensor" />
        <DataSource Type="Tool" Id="FilteredTiltSensor" PortName="FilteredTiltSensor" />
        <DataSource Type="Tool" Id="Accelerometer" PortName="Accelerometer" />
        <DataSource Type="Tool" Id="Gyroscope" PortName="Gyroscope" />
        <DataSource Type="Tool" Id="Magnetometer" PortName="Magnetometer" />
      </DataSources>
      <OutputChannels>
        <OutputChannel Id="TrackerStream" >
          <DataSource Id="OrientationSensor"/>
          <DataSource Id="TiltSensor"/>
          <DataSource Id="FilteredTiltSensor"/>
          <DataSource Id="Accelerometer"/>
          <DataSource Id="Gyroscope"/>
          <DataSource Id="Magnetometer"/>
        </OutputChannel>
      </OutputChannels>
    </Device>
    <Device
      Id="CaptureDevice"
      Type="VirtualCapture"
      BaseFilename="RecordingTest.igs.mha"
      EnableCapturingOnStart="FALSE" >
      <InputChannels>
        <InputChannel Id="TrackerStream" />
      </InputChannels>
    </Device>    
  </DataCollection>
  
  <PlusOpenIGTLinkServer 
    MaxNumberOfIgtlMessagesToSend="1" 
    MaxTimeSpentWithProcessingMs="50" 
    ListeningPort="18944" 
    OutputChannelId="TrackerStream" > 
    <DefaultClientInfo> 
      <MessageTypes> 
        <Message Type="TRANSFORM" />
      </MessageTypes>
      <TransformNames> 
        <Transform Name="OrientationSensorToTracker" /> 
        <Transform Name="TiltSensorToTracker" /> 
        <Transform Name="FilteredTiltSensorToTracker" /> 
        <xTransform Name="AccelerometerToTracker" /> 
        <xTransform Name="GyroscopeToTracker" /> 
        <xTransform Name="MagnetometerToTracker" /> 
      </TransformNames>
    </DefaultClientInfo>
  </PlusOpenIGTLinkServer>
  
</PlusConfiguration>
````

**Comment by LisanneVenix** [comment]
This was in the log file: 
|458.363000|SERVER> Device TrackerDevice cannot connect to PhidgetSpatial hardware device (serial number: any). Problem waiting for attachment (Timed Out)| in :\D\PSNPb\PlusLib\src\PlusDataCollection\PhidgetSpatialTracking\vtkPlusPhidgetSpatialTracker.cxx(382)
|ERROR|458.392000|SERVER> TrackerDevice: Cannot connect to data source, ConnectInternal failed| in :\D\PSNPb\PlusLib\src\PlusDataCollection\vtkPlusDevice.cxx(1147)
|ERROR|458.393000|SERVER> Unable to connect device: TrackerDevice.| in :\D\PSNPb\PlusLib\src\PlusDataCollection\vtkPlusDataCollector.cxx(353)
|ERROR|458.394000|SERVER> Datacollector failed to connect to devices| in :\D\PSNPb\PlusLib\src\PlusServer\Tools\PlusServer.cxx(106)
|ERROR|458.679000| Server stopped unexpectedly. Return code: 1| in E:\D\PSNPb\PlusApp\PlusServerLauncher\PlusServerLauncherMainWindow.cxx(864)
|INFO|458.680000| Disconnect request successful| in E:\D\PSNPb\PlusApp\PlusServerLauncher\PlusServerLauncherMainWindow.cxx(632)

**Reply by adamrankin** [reply]
Could you paste the section of your config file for your phidget device?

**Reply by LisanneVenix** [reply]
````
<PlusConfiguration version="2.1">
  <DataCollection StartupDelaySec="1.0">
    <DeviceSet 
      Name="PlusServer: PhidgetSpatial MARG sensor" 
      Description="Broadcasting tool tracking data through OpenIGTLink
Keep the device stationary for 2 seconds after connect for accurate gyroscope initialization."
    />
    <Device
      Id="TrackerDevice"
      Type="PhidgetSpatial"
      AhrsAlgorithm="MADGWICK_MARG"
      AhrsAlgorithmGain="1.5"
      TiltSensorWestAxisIndex="1"
      FilteredTiltSensorAhrsAlgorithm="MADGWICK_IMU"
      FilteredTiltSensorAhrsAlgorithmGain="0.01"
      FilteredTiltSensorWestAxisIndex="1"      
      ToolReferenceFrame="Tracker" >
      <DataSources>
        <DataSource Type="Tool" Id="OrientationSensor" PortName="OrientationSensor" />    
        <DataSource Type="Tool" Id="TiltSensor" PortName="TiltSensor" />
        <DataSource Type="Tool" Id="FilteredTiltSensor" PortName="FilteredTiltSensor" />
        <DataSource Type="Tool" Id="Accelerometer" PortName="Accelerometer" />
        <DataSource Type="Tool" Id="Gyroscope" PortName="Gyroscope" />
        <DataSource Type="Tool" Id="Magnetometer" PortName="Magnetometer" />
      </DataSources>
      <OutputChannels>
        <OutputChannel Id="TrackerStream" >
          <DataSource Id="OrientationSensor"/>
          <DataSource Id="TiltSensor"/>
          <DataSource Id="FilteredTiltSensor"/>
          <DataSource Id="Accelerometer"/>
          <DataSource Id="Gyroscope"/>
          <DataSource Id="Magnetometer"/>
        </OutputChannel>
      </OutputChannels>
    </Device>
    <Device
      Id="CaptureDevice"
      Type="VirtualCapture"
      BaseFilename="RecordingTest.igs.mha"
      EnableCapturingOnStart="FALSE" >
      <InputChannels>
        <InputChannel Id="TrackerStream" />
      </InputChannels>
    </Device>    
  </DataCollection>
  
  <PlusOpenIGTLinkServer 
    MaxNumberOfIgtlMessagesToSend="1" 
    MaxTimeSpentWithProcessingMs="50" 
    ListeningPort="18944" 
    OutputChannelId="TrackerStream" > 
    <DefaultClientInfo> 
      <MessageTypes> 
        <Message Type="TRANSFORM" />
      </MessageTypes>
      <TransformNames> 
        <Transform Name="OrientationSensorToTracker" /> 
        <Transform Name="TiltSensorToTracker" /> 
        <Transform Name="FilteredTiltSensorToTracker" /> 
        <xTransform Name="AccelerometerToTracker" /> 
        <xTransform Name="GyroscopeToTracker" /> 
        <xTransform Name="MagnetometerToTracker" /> 
      </TransformNames>
    </DefaultClientInfo>
  </PlusOpenIGTLinkServer>
  
</PlusConfiguration>
````

**Reply by LisanneVenix** [reply]
http://perk-software.cs.queensu.ca/plus/doc/nightly/user/DevicePhidgetSpatial.html

**Reply by adamrankin** [reply]
Does Phidget have its own software so you can confirm the device is working first?

**Reply by LisanneVenix** [reply]
Yes, it does. It is working

**Reply by adamrankin** [reply]
Just in case, could you try again ensuring that the Phidget software is not running?

**Reply by LisanneVenix** [reply]
Yes, it is not running

**Reply by LisanneVenix** [reply]
|DEBUG|1986.007000|SERVER> StartupDelaySec: 1.000000| in :\D\PSNPb\PlusLib\src\PlusDataCollection\vtkPlusDataCollector.cxx(89)
|DEBUG|1986.009000|SERVER> Buffer size is not defined in source element "MagnetometerToTracker". Using default buffer size: 150| in :\D\PSNPb\PlusLib\src\PlusDataCollection\vtkPlusDataSource.cxx(392)
|DEBUG|1986.010000|SERVER> AveragedItemsForFiltering is not defined in source element "MagnetometerToTracker". Using default value: 20| in :\D\PSNPb\PlusLib\src\PlusDataCollection\vtkPlusDataSource.cxx(402)
|DEBUG|1986.010000|SERVER> TrackerDevice: Unable to find acquisition rate in device element when it is required, using default 125| in :\D\PSNPb\PlusLib\src\PlusDataCollection\vtkPlusDevice.cxx(1000)
|DEBUG|1986.010000|SERVER> TrackerDevice: Local time offset was not defined in device configuration| in :\D\PSNPb\PlusLib\src\PlusDataCollection\vtkPlusDevice.cxx(1053)
|DEBUG|1986.011000|SERVER> CaptureDevice: Unable to find acquisition rate in device element when it is required, using default 30| in :\D\PSNPb\PlusLib\src\PlusDataCollection\vtkPlusDevice.cxx(1000)
|DEBUG|1986.011000|SERVER> CaptureDevice: Local time offset was not defined in device configuration| in :\D\PSNPb\PlusLib\src\PlusDataCollection\vtkPlusDevice.cxx(1053)
|INFO|1986.012000|SERVER> Server status: Reading configuration.| in :\D\PSNPb\PlusLib\src\PlusServer\Tools\PlusServer.cxx(92)
|DEBUG|1986.013000|SERVER> vtkPlusTransformRepository::ReadConfiguration: no CoordinateDefinitions element was found| in :\D\PSNPb\PlusLib\src\PlusCommon\vtkPlusTransformRepository.cxx(625)
|INFO|1986.013000|SERVER> Server status: Connecting to devices.| in :\D\PSNPb\PlusLib\src\PlusServer\Tools\PlusServer.cxx(103)
|DEBUG|1986.014000|SERVER> TrackerDevice: vtkPlusDevice::Connect: TrackerDevice| in :\D\PSNPb\PlusLib\src\PlusDataCollection\vtkPlusDevice.cxx(1134)
|DEBUG|1986.014000|SERVER> Waiting for phidget spatial device to be attached...| in :\D\PSNPb\PlusLib\src\PlusDataCollection\PhidgetSpatialTracking\vtkPlusPhidgetSpatialTracker.cxx(375)
|ERROR|1995.994000|SERVER> Device TrackerDevice cannot connect to PhidgetSpatial hardware device (serial number: any). Problem waiting for attachment (Timed Out)| in :\D\PSNPb\PlusLib\src\PlusDataCollection\PhidgetSpatialTracking\vtkPlusPhidgetSpatialTracker.cxx(382)
|ERROR|1996.047000|SERVER> TrackerDevice: Cannot connect to data source, ConnectInternal failed| in :\D\PSNPb\PlusLib\src\PlusDataCollection\vtkPlusDevice.cxx(1147)
|ERROR|1996.047000|SERVER> Unable to connect device: TrackerDevice.| in :\D\PSNPb\PlusLib\src\PlusDataCollection\vtkPlusDataCollector.cxx(353)
|ERROR|1996.048000|SERVER> Datacollector failed to connect to devices| in :\D\PSNPb\PlusLib\src\PlusServer\Tools\PlusServer.cxx(106)
|DEBUG|1996.049000|SERVER> CaptureDevice: vtkPlusDevice::Connect: CaptureDevice| in :\D\PSNPb\PlusLib\src\PlusDataCollection\vtkPlusDevice.cxx(1134)
|DEBUG|1996.050000|SERVER> TrackerDevice: vtkPlusDevice::Disconnect| in :\D\PSNPb\PlusLib\src\PlusDataCollection\vtkPlusDevice.cxx(1159)
|DEBUG|1996.050000|SERVER> TrackerDevice: Data source is already disconnected| in :\D\PSNPb\PlusLib\src\PlusDataCollection\vtkPlusDevice.cxx(1163)
|DEBUG|1996.051000|SERVER> CaptureDevice: vtkPlusDevice::Disconnect| in :\D\PSNPb\PlusLib\src\PlusDataCollection\vtkPlusDevice.cxx(1159)
|DEBUG|1996.051000|SERVER> vtkPlusDataCollector::Disconnect: All devices have been disconnected| in :\D\PSNPb\PlusLib\src\PlusDataCollection\vtkPlusDataCollector.cxx(393)
|DEBUG|1996.052000|SERVER> No saved data source devices were found that use original timestamps, so synchronization of loop times is not performed| in :\D\PSNPb\PlusLib\src\PlusDataCollection\vtkPlusDataCollector.cxx(674)
|ERROR|1996.147000| Server stopped unexpectedly. Return code: 1| in E:\D\PSNPb\PlusApp\PlusServerLauncher\PlusServerLauncherMainWindow.cxx(864)
|INFO|1996.149000| Disconnect request successful| in E:\D\PSNPb\PlusApp\PlusServerLauncher\PlusServerLauncherMainWindow.cxx(632)

**Reply by adamrankin** [reply]
I'm not very familiar with the Phidget device. Can you click connect, then try unplugging and plugging the device while it's trying to connect?

**Reply by LisanneVenix** [reply]
It is giving the exact same error! By the way, I am not pressing 'connect' but i am pressing the 'Launch server' button. Does that mean I am doing something wrong?

**Reply by adamrankin** [reply]
No, sorry, I just used the wrong words. I'm not sure what to do. It is failing here:
https://github.com/PlusToolkit/PlusLib/blob/master/src/PlusDataCollection/PhidgetSpatialTracking/vtkPlusPhidgetSpatialTracker.cxx#L376

but I don't know why. @lassoan @Sunderlandkyl any experience with the Phidget tracker?

**Reply by Sunderlandkyl** [reply]
Sorry, I don't have much experience with it.

**Reply by dgmato** [reply]
@LisanneVenix if you've recently purchased the Phidget device, you might have acquired the model [MOT0109_0](https://www.phidgets.com/?tier=3&catid=10&pcid=8&prodid=1204). This new model uses a different processor and it is not compatible with the old library. See message in website: 

> Due to parts and components going obsolete and becoming unavailable, the MOT0109_0 uses a different processor than the 1044_1B that is not supported under Phidget21. If you're using our old libraries you must [upgrade](https://www.phidgets.com/docs/Upgrading_Code_from_Phidget21_to_Phidget22) to Phidget22 before you can make use of this Phidget.

It seems that in order to use this new model, Phidget libraries should be upgraded in PLUS toolkit.

I hope this helps.
